rust/worker/src/compactor/scheduler.rs

```diff
@@ -0,0 +1,397 @@
+                .await;
+
+            match result {
+                Ok(collection) => {
+                    if collection.is_empty() {
+                        // TODO: Log error
+                        println!("Collection not found: {:?}", collection_info.collection_id);
+                        continue;
+                    }
+                    collection_records.push(CollectionRecord {
+                        id: collection[0].id.to_string(),
+                        tenant_id: collection[0].tenant.clone(),
+                        // TODO: get the last compaction time from the sysdb
+                        last_compaction_time: 0,
+                        first_record_time: collection_info.first_log_id_ts,
+                        cursor: collection_info.first_log_id,
+                    });
+                }
+                Err(e) => {
+                    // TODO: Log error
+                    println!("Error: {:?}", e);
+                }
+            }
+        }
+        collection_records
+    }
+
+    pub(crate) async fn schedule_internal(&mut self, collection_records: Vec<CollectionRecord>) {
+        let tasks = self
+            .policy
+            .determine(collection_records, self.max_queue_size as i32);
+        {
+            let mut task_queue = self.task_queue.lock();
+            task_queue.clear();
+            task_queue.extend(tasks);
+        }
+    }
+
+    pub(crate) async fn schedule(&mut self) {
+        let collections = self.get_collections_with_new_data().await;
+        if collections.is_empty() {
+            return;
+        }
+        let collection_records = self.verify_and_enrich_collections(collections).await;
+        self.schedule_internal(collection_records).await;
+    }
+
+    pub(crate) fn take_task(&self) -> Option<Task> {
+        let mut task_queue = self.task_queue.lock();
+        if task_queue.is_empty() {
+            return None;
+        }
+        Some(task_queue.remove(0))
+    }
+
+    pub(crate) fn get_tasks(&self) -> Vec<Task> {
+        let task_queue = self.task_queue.lock();
+        task_queue.clone()
+    }
+}
+
+impl Component for Scheduler {
+    fn on_start(&mut self, ctx: &ComponentContext<Self>) {
+        ctx.scheduler.schedule_interval(
+            ctx.sender.clone(),
+            ScheduleMessage {},
+            self.schedule_interval,
+            None,
+            ctx,
+        );
+    }
+
+    fn queue_size(&self) -> usize {
+        1000
+    }
+}
+
+impl Debug for Scheduler {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Scheduler")
+    }
+}
+
+#[derive(Clone, Debug)]
+struct ScheduleMessage {}
+
+#[async_trait]
+impl Handler<ScheduleMessage> for Scheduler {
+    async fn handle(&mut self, _event: ScheduleMessage, _ctx: &ComponentContext<Scheduler>) {
+        self.schedule().await;
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct SchedulerPolicy {}
+
+impl SchedulerPolicy {
```

Comment:
Should this use a trait and should the scheduler be generic over a Box<dyn SchedulerPolicy>?